Wrap NULL/0 is-a-null-pointer in a utility function

tautschnig opened the pull request:
A pointer-typed constant expression with value "NULL" is always a null
pointer expression. When NULL_is_zero is true in the platform
configuration, a value "0" is also a null pointer.

These possible options were repeatedly, but not consistently, evaluated
ad-hoc. This resulted in a least one missed simplification opportunity
as we noticed on an RMC-generated goto binary (which had 0 instead of
NULL in a pointer-typed constant).

<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.

If your PR fixes a bug, include the regression test(s) in the same commit as the bug fix. Else, keep commits small and orthogonal, possibly placing tests in commits of their own.
--->

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [x] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

tautschnig committed:
Wrap NULL/0 is-a-null-pointer in a utility function

A pointer-typed constant expression with value "NULL" is always a null
pointer expression. When NULL_is_zero is true in the platform
configuration, a value "0" is also a null pointer.

These possible options were repeatedly, but not consistently, evaluated
ad-hoc. This resulted in a least one missed simplification opportunity
as we noticed on an RMC-generated goto binary (which had 0 instead of
NULL in a pointer-typed constant).

NlightNFotis commented on src/util/simplify_expr_int.cpp:
```diff
@@ -1572,9 +1572,12 @@ simplify_exprt::resultt<> simplify_exprt::simplify_inequality_rhs_is_constant(
+    if(expr.id() != ID_equal)
+      return unchanged(expr);
+
+    const constant_exprt &op1_constant = to_constant_expr(expr.op1());
```
Is it possible that this transfer function throws? Say, if `expr.op1()` is not a `constant_exprt`?

Before this change, a property was checked and there was no transfer function application, so I'm slightly worried that this could result in invariants at runtime in some cases.

tautschnig replied:
This previous check actually was unnecessary: the function checks for this earlier on already.

NlightNFotis commented on src/goto-instrument/dump_c.cpp:
```diff
@@ -1380,8 +1379,7 @@ void dump_ct::cleanup_expr(exprt &expr)
             // add a typecast for NULL or 0
             if(
               argument.id() == ID_constant &&
```
Could this also be refactored to `argument.is_constant()`? Or is this not possible?

tautschnig replied:
Yes, which should likely be cleaned up across the code base. May I defer this to a PR of its own?

NlightNFotis replied:
Yes, no problem.

tautschnig replied:
Now done in https://github.com/diffblue/cbmc/pull/6542.

NlightNFotis reviewed (approved):
Solid, minor questions/comments from my end.

tautschnig committed:
Disable support for constant(zero) as NULL

All front-ends and any other code should take care of generating
constant(NULL) instead.
